### PR TITLE
Vectorize non-persistent Softmax

### PR DESCRIPTION
**Pull request by slayton58**
Resubmit of #36485 with bug fix & enhanced testing.

Moved `test_softmax_backward` -> `test_softmax_results`, check fprop & bgrad against CPU implementation for all cases.

\cc @ngimel

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -9726,20 +9726,35 @@ def test_embedding_padding_idx(self, device, dtype):
...
+                    # Note: With the largest tests we can hit upper limit of fp16 when we
+                    # sum, so scale the input down to stay in a nicer range.
+                    if dtype == torch.float16:
+                        input = input / 100.
                     input = input[shift[0]:, shift[1]:]
+                    # Note; Don't want to bprop back through slice op
+                    input = input.detach().requires_grad_(True)
+                    ref_input = input.clone().cpu().detach().requires_grad_(True)
                     for dim in [0, 1]:
+                        ref_output = fn(ref_input, dtype=torch.float, dim=dim).sum()
```
Comment (ngimel) [inline test/test_nn.py:9747]: nit: when you calculate output as .sum(), gradOutput will be filled with ones, which won't let you detect if you are reading gradOutput values from the wrong places. You can do
``` 
output=fn(input)
gradOutput = torch.rand_like(output)
grad_input, = torch.autograd.grad(output, input, grad_outputs=(gradOutput), create_graph=True)
```

File: test/test_nn.py [hunk 2]
```diff
@@ -9726,20 +9726,35 @@ def test_embedding_padding_idx(self, device, dtype):
...
                         output = fn(input, dtype=torch.float, dim=dim).sum()
                         grad_input, = torch.autograd.grad(output, input, create_graph=True)
+                        ref_grad_input, = torch.autograd.grad(ref_output, ref_input, create_graph=True)
                         grad_input.sum().backward()
+                        ref_grad_input.sum().backward()
+
+                        self.assertEqual(output, ref_output)
```
Comment (ngimel) [inline test/test_nn.py:9754]: can you also assert grad_input and ref_grad_input?

File: test/test_nn.py (hunk 1 above)
Comment (slayton58) [reply test/test_nn.py:9747]: Is that style necessary or is this syntax acceptable?
```
grad_output = torch.rand_like(output)
output.backward(grad_output)
```

File: test/test_nn.py [hunk 3]
```diff
@@ -9726,20 +9726,35 @@ def test_embedding_padding_idx(self, device, dtype):
 
     @dtypesIfCUDA(torch.half, torch.float)
     @dtypes(torch.float)
-    def test_softmax_backward(self, device, dtype):
+    def test_softmax_results(self, device, dtype):
         # Non-even sizes and non-zero shifts test fallback paths in vectorized kernel
         # Note: dim1 > 1024 is needed to exercise the vectorized (non-persistent) path, (16, 30576) is BERT-esque
         sizes = [(0, 10), (32, 20), (10, 0), (31, 20), (32, 21), (31, 23), (32, 1536), (31, 2048), (16, 30576)]
```
[the remark this answers is not in the document]
Comment (slayton58) [reply test/test_nn.py:9732]: Done - added `(33, 2049)`

File: test/test_nn.py (hunk 1 above)
Comment (slayton58) [reply test/test_nn.py:9747]: Ended up using your syntax (although I'm still interested in _why_ that syntax over the simpler one I mentioned)

File: test/test_nn.py (hunk 2 above)
Comment (slayton58) [reply test/test_nn.py:9754]: Done
